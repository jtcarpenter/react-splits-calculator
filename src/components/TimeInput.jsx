import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledLabel = styled.label`
    border: 1px solid #777777;
    padding: 15px;
    border-radius: 5px;
`;

const StyledInput = styled.input`
    border: none;
    border-radius: 3px;
    padding: 10px 5px;
    text-align: center;
    width: 1.5em; // hour input will be less
    outline: none;
`;

export const HOUR_ID = 'input-hour';
export const MINUTE_ID = 'input-minute';
export const SECOND_ID = 'input-second';

const VALID_REGEXP = {
    [HOUR_ID]: /^$|^\d$/,
    [MINUTE_ID]: /^([0-5]\d?)?$/,
    [SECOND_ID]: /^([0-5]\d?)?$/
};

const KEY_CODE_BACKSPACE = 8;

export class TimeInput extends PureComponent {

    constructor(props) {
        super(props);
        this.props = props;

        this.state = {
            [HOUR_ID]: '',
            [MINUTE_ID]: '',
            [SECOND_ID]: ''
        };

        this.handleKeyDown = this.handleKeyDown.bind(this);
        this.handleKeyPress = this.handleKeyPress.bind(this);
        this.handleChange = this.handleChange.bind(this);

        this.inputs = [
            {
                id: HOUR_ID,
                next: MINUTE_ID,
                el: null,
                placeholder: '0',
                maxLength: 1
            },
            {
                id: MINUTE_ID,
                prev: HOUR_ID,
                next: SECOND_ID,
                el: null,
                placeholder: '00',
                maxLength: 2
            },
            {
                id: SECOND_ID,
                prev: MINUTE_ID,
                el: null,
                placeholder: '00',
                maxLength: 2
            }
        ];
    }

    handleKeyDown(evt) {
        const { keyCode, target } = evt;
        const { dataset = {}, id } = target;
        const { prev } = dataset;
        if (prev && keyCode === KEY_CODE_BACKSPACE) {
            if (this.state[id].length === 0) {
                this.inputs.find((input) => input.id === prev).el.focus();
                evt.preventDefault();
            }
        }
    }

    handleKeyPress(evt) {
        const { target } = evt;
        const { dataset = {}, selectionStart } = target;
        const { next, maxLength } = dataset;
        if (next && selectionStart === maxLength) {
            this.inputs.find((input) => input.id === next).el.focus();
        }
    }

    handleChange(evt) {
        const { target } = evt;
        const { dataset = {}, id, value } = target;
        const { next, maxLength, prev } = dataset;
        const valid = VALID_REGEXP[id].test(value);
        if (!valid) {
            evt.preventDefault();
            return;
        }
        this.setState({ [id]: value });
        if (next && value.length === parseInt(maxLength, 10)) {
            this.inputs.find((input) => input.id === next).el.focus();
        }
        if (prev && value.length === 0) {
            this.inputs.find((input) => input.id === prev).el.focus();
        }
        this.props.onChange({
            hour: id === HOUR_ID ? value : this.state[HOUR_ID],
            minute: id === MINUTE_ID ? value : this.state[MINUTE_ID],
            second: id === SECOND_ID ? value : this.state[SECOND_ID]
        });
    }

    render() {
        return (
            <StyledLabel>
                time:
                {this.inputs.map((input, index) => 
                    <StyledInput
                        key={ index }
                        innerRef={ (el) => { input.el = el; } }
                        type="string"
                        onKeyDown={ this.handleKeyDown }
                        onKeyPress={ this.handleKeyPress }
                        onChange={ this.handleChange }
                        placeholder={ input.placeholder }
                        value={ this.state[input.id] }
                        name={ input.id }
                        id={ input.id }
                        data-prev={ input.prev }
                        data-next={ input.next }
                        data-max-length={ input.maxLength }
                    />
                )}
            </StyledLabel>
        );
    }
}

TimeInput.propTypes = {
    onChange: PropTypes.func.isRequired
};

export default TimeInput;