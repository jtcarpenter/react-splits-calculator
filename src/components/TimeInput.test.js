import React from 'react';
import { shallow, mount } from 'enzyme';
import { TimeInput } from './TimeInput';

import renderer from 'react-test-renderer';

const timeInput = (
    <TimeInput onChange={() => {}}/>
);

describe('TimeInput', () => {

    let wrapper;
    let shallowWrapper;
    let hourInput;
    let minuteInput;
    let secondInput;

    beforeEach(() => {
        wrapper = mount(timeInput);
        shallowWrapper = shallow(timeInput);
        hourInput = wrapper.find('input#hour');
        minuteInput = wrapper.find('input#minute');
        secondInput = wrapper.find('input#second');
    });

    afterEach(() => {
        document.activeElement.blur();
    });

    it('should render', () => {
        expect(shallowWrapper.exists()).toBe(true);
    });

    it('should display correctly', () => {
        expect(renderer.create(timeInput).toJSON()).toMatchSnapshot();
    });

    it('should update hour state to 1 when input changes to 1', () => {
        expect(wrapper.state('hour')).toBe('');
        hourInput.simulate('change', { target: { value: '1' } });
        expect(wrapper.state('hour')).toBe('1');
    });

    it('should move focus to minutes when hour is 1 character long', () => {
        hourInput.simulate('change', { target: { value: '1' } });
        const focusedElement = document.activeElement;
        expect(focusedElement.id).toEqual(minuteInput.props().id);
    });

    it('should move focus to seconds when miutes is 2 characters long', () => {
        minuteInput.simulate('change', { target: { value: '11' } });
        const focusedElement = document.activeElement;
        expect(focusedElement.id).toEqual(secondInput.props().id);
    });

    it('should update seconds state to empty after deleting', () => {
        secondInput.simulate('change', { target: { value: '1' } });
        expect(wrapper.state('second')).toBe('1');
        secondInput.simulate('change', { target: { value: '' } });
        expect(wrapper.state('second')).toBe('');
    });

    it('should update minute state to empty after deleting', () => {
        minuteInput.simulate('change', { target: { value: '1' } });
        expect(wrapper.state('minute')).toBe('1');
        minuteInput.simulate('change', { target: { value: '' } });
        expect(wrapper.state('minute')).toBe('');
    });

    it('should update hour state to empty after deleting', () => {
        hourInput.simulate('change', { target: { value: '1' } });
        expect(wrapper.state('hour')).toBe('1');
        hourInput.simulate('change', { target: { value: '' } });
        expect(wrapper.state('hour')).toBe('');
    });

    it('should move focus to minutes if seconds is empty after delete', () => {
        secondInput.simulate('change', { target: { value: '1' } });
        expect(wrapper.state('second')).toBe('1');
        secondInput.simulate('change', { target: { value: '' } });
        const focusedElement = document.activeElement;
        expect(focusedElement.id).toEqual(minuteInput.props().id);
    });

    it('should move focus to hours if minutes is empty after delete', () => {
        minuteInput.simulate('change', { target: { value: '1' } });
        expect(wrapper.state('minute')).toBe('1');
        minuteInput.simulate('change', { target: { value: '' } });
        const focusedElement = document.activeElement;
        expect(focusedElement.id).toEqual(hourInput.props().id);
    });
});