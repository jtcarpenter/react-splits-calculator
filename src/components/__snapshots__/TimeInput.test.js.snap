// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeInput should display correctly 1`] = `
<label
  className="sc-bdVaJa eKxwv"
>
  time:
  <input
    className="sc-bwzfXH gEvcfP"
    data-max-length={1}
    data-next="input-minute"
    id="input-hour"
    name="input-hour"
    onChange={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    placeholder="0"
    type="string"
    value=""
  />
  <input
    className="sc-bwzfXH gEvcfP"
    data-max-length={2}
    data-next="input-second"
    data-prev="input-hour"
    id="input-minute"
    name="input-minute"
    onChange={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    placeholder="00"
    type="string"
    value=""
  />
  <input
    className="sc-bwzfXH gEvcfP"
    data-max-length={2}
    data-prev="input-minute"
    id="input-second"
    name="input-second"
    onChange={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    placeholder="00"
    type="string"
    value=""
  />
</label>
`;
